File: Companies/Google/UniquePaths.java
package Companies.Google;

import java.util.Arrays;

public class UniquePaths {
    public int uniquePaths(int m, int n) {
        int[] dp = new int[n];
        Arrays.fill(dp, 1);
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                dp[j] += dp[j-1];
            }
        }
        return dp[n-1];
    }

    /* Obstacles */
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.length, n = obstacleGrid[0].length;
        int[] dp = new int[n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (obstacleGrid[i][j] == 1) {
                    dp[j] = 0;
                } else if (j == 0) {
                    if (i == 0) {
                        dp[j] = 1;
                    } else {
                        continue;
                    }
                } else {
                    dp[j] += dp[j-1];
                }
            }
        }
        return dp[n-1];
    }

    private int sx, sy, ex, ey, aval = 0, re = 0;
    private int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};

    public int uniquePathsIII(int[][] grid) {
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] >= 0) {
                    aval += 1;
                    if (grid[i][j] == 1) {
                        sx = i;
                        sy = j;
                    } else if (grid[i][j] == 2) {
                        ex = i;
                        ey = j;
                    }
                }
            }
        }
        dfs(grid, sx, sy, aval);
        return re;
    }

    private void dfs(int[][] grid, int x, int y, int aval) {
        if (x == ex && y == ey) {
            if (aval == 1) {
                re++;
                return;
            }
        }
        grid[x][y] = -2;
        for (int[] dir : dirs) {
            int xx = x + dir[0];
            int yy = y + dir[1];
            if (xx >= 0 && xx < grid.length && yy >= 0 && yy < grid[0].length && grid[xx][yy] >= 0) {
                dfs(grid, xx, yy, aval-1);
            }
        }
        grid[x][y] = 0;
    }
}

File: Topics/Design/al/Unique_Paths.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<int> dp(n, 1);
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                dp[j] += dp[j - 1];
            }
        }
        return dp[n - 1];
    }

    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        vector<int> dp(n, 0);

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (obstacleGrid[i][j] == 1) {
                    dp[j] = 0;
                }
                else if (j == 0) {
                    if (i == 0)
                        dp[j] = 1;
                }
                else {
                    dp[j] += dp[j - 1];
                }
            }
        }
        return dp[n - 1];
    }

    int uniquePathsIII(vector<vector<int>>& grid) {
        rows = grid.size();
        cols = grid[0].size();

        available = 0;
        result = 0;

        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                if (grid[i][j] >= 0) {
                    ++available;
                    if (grid[i][j] == 1) {
                        startX = i;
                        startY = j;
                    }
                    else if (grid[i][j] == 2) {
                        endX = i;
                        endY = j;
                    }
                }
            }
        }

        dfs(grid, startX, startY, available);
        return result;
    }

private:
    int rows, cols;

    int startX, startY;
    int endX, endY;

    int available;
    int result;

    vector<pair<int, int>> dirs{
        {1, 0},
        {-1, 0},
        {0, 1},
        {0, -1}
    };

    void dfs(vector<vector<int>>& grid,
             int x,
             int y,
             int remain) {

        if (x == endX && y == endY) {
            if (remain == 1) {
                ++result;
            }
            return;
        }

        grid[x][y] = -2;      // visited

        for (auto [dx, dy] : dirs) {
            int nx = x + dx;
            int ny = y + dy;

            if (nx >= 0 && nx < rows &&
                ny >= 0 && ny < cols &&
                grid[nx][ny] >= 0) {

                dfs(grid, nx, ny, remain - 1);
            }
        }

        grid[x][y] = 0;       // backtrack
    }
};
